modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxUnlockManager.java

```diff
@@ -0,0 +1,321 @@
+                                        lockReleaseMsg.txId(),
+                                        groupId,
+                                        ATTEMPTS_TO_CLEANUP_REPLICA
+                                );
+                            }
+                        });
+                    }
+
+                    clusterService.messagingService().respond(senderId, msg, correlationId);
+                });
+    }
+
+    // TODO https://issues.apache.org/jira/browse/IGNITE-20681 remove attempts count.
+    private CompletableFuture<Void> cleanupWithRetry(
+            boolean commit,
+            @Nullable HybridTimestamp commitTimestamp,
+            UUID txId,
+            TablePartitionId partitionId,
+            int attempts) {
+        return findPrimaryReplica(partitionId, hybridClock.now())
+                .thenCompose(leaseHolder ->
+                        txManager.cleanup(leaseHolder, partitionId, txId, commit, commitTimestamp))
+                .handle((res, ex) -> {
+                    if (ex != null) {
+                        if (attempts > 0) {
+                            LOG.warn("Failed to perform cleanup on Tx. The operation will be retried [txId={}].", txId, ex);
+                        } else {
+                            LOG.warn("Failed to perform cleanup on Tx [txId={}].", txId, ex);
+                        }
+
+                        if (attempts > 0) {
+                            return cleanupWithRetry(commit, commitTimestamp, txId, partitionId, attempts - 1);
+                        }
+
+                        return CompletableFuture.<Void>failedFuture(ex);
+                    }
+
+                    return CompletableFuture.<Void>completedFuture(null);
+                })
+                .thenCompose(Function.identity());
+    }
+
+    private CompletableFuture<ReplicaMeta> getPrimaryReplica(TablePartitionId partitionId) {
+        return placementDriver.getPrimaryReplica(partitionId, hybridClock.now());
+    }
+
+    private CompletableFuture<String> findPrimaryReplica(TablePartitionId partitionId, HybridTimestamp now) {
+        return placementDriver.awaitPrimaryReplica(partitionId, now, AWAIT_PRIMARY_REPLICA_TIMEOUT, SECONDS)
+                .handle((primaryReplica, e) -> {
+                    if (e != null) {
+                        LOG.error("Failed to retrieve primary replica for partition {}", partitionId, e);
+
+                        throw withCause(TransactionException::new, REPLICA_UNAVAILABLE_ERR,
+                                "Failed to get the primary replica"
+                                        + " [tablePartitionId=" + partitionId + ", awaitTimestamp=" + now + ']', e);
+                    }
+
+                    return primaryReplica.getLeaseholder();
+                });
+    }
+
+    @Override
+    public void stop() throws Exception {
+    }
+
+    private CompletableFuture<Void> sendDurableUnlockMessage(
```

Comment:
Why is it called `durable`? I see that locks are marked as released after unlock, I just mean that this should be something like `sendUnlockMessageWithRetries`